lib/terrafying/generator.rb

```diff
@@ -9,20 +9,52 @@ module Terrafying
 
   class Ref
 
-    def initialize(var)
-      @var = var
+    def initialize(
+          kind: :resource,
+          type: "",
+          name:,
+          key: nil,
+          fns: []
+        )
+      @kind = kind
+      @type = type
+      @name = name
+      @key = key
+      @fns = fns
+    end
+
+    def fn_call(fn)
+      Ref.new(kind: @kind, type: @type, name: @name, key: @key, fns: [fn] + @fns)
     end
 
     def downcase
-      Ref.new("lower(#{@var})")
+      fn_call("lower")
     end
 
     def strip
-      Ref.new("trimspace(#{@var})")
+      fn_call("trimspace")
     end
 
     def to_s
-      "${#{@var}}"
+      closing_parens = ")" * @fns.count
+      calls = ""
+      if @fns.count > 0
+        calls = @fns.join("(") + "("
+      end
+
+      type = @type
+      if @kind != :resource
+        type = [@kind.to_s, @type]
+      end
+
+      key = @key
+      if @kind == :resource && key == nil
+        key = "id"
+      end
```

Comment:
We can reuse out `resource?` methods here for some fun:

```ruby
key = @key
resource? && key ||= 'id'
```

Reply:
I don't feel like drop in clarity is worth the drop of chars